source/Nuke.Common/Tools/OctoVersion/OctoVersionAttribute.cs

```diff
@@ -22,27 +24,232 @@ namespace Nuke.Common.Tools.OctoVersion
+            Logger.Info("Resolved OctoVersion Properties");
+            Logger.Info($"  Framework:                 {Framework ?? "null"}");
+            Logger.Info($"  UpdateBuildNumber:         {UpdateBuildNumber}");
+            Logger.Info($"  AutoDetectBranch:          {octoVersionAutoDetectBranch?.ToString() ?? "null"}");
+            Logger.Info($"  Branch:                    {octoversionBranch ?? "null"}");
+            Logger.Info($"  Major:                     {octoVersionMajor?.ToString() ?? "null"}");
+            Logger.Info($"  Minor:                     {octoVersionMinor?.ToString() ?? "null"}");
+            Logger.Info($"  Patch:                     {octoVersionPatch?.ToString() ?? "null"}");
+
+            // Auto Detect branch
+            if (octoVersionAutoDetectBranch.HasValue && octoVersionAutoDetectBranch.Value)
+            {
+                // Don't allow auto detect if a Branch was set manually via property or parameter
+                ControlFlow.Assert(string.IsNullOrEmpty(octoversionBranch), $"If {nameof(AutoDetectBranch)} is enabled, then a branch can not be specified via {nameof(Branch)} or {nameof(BranchParameter)}.");
+
+                Logger.Info($"Detecting current branch from the local git directory.");
+                var gitRepository = GitRepository.FromLocalDirectory(NukeBuild.RootDirectory);
+                octoversionBranch = gitRepository.Branch;
+                Logger.Info($"Current branch is '{octoversionBranch}'.");
+            }
+
+            if (string.IsNullOrEmpty(octoversionBranch))
+            {
+                // If we have ended up with no branch, abort!
+                ControlFlow.Fail($"OctoVersion requires the branch to be specified.  Either enable auto detection via {nameof(AutoDetectBranch)} or {nameof(AutoDetectBranchParameter)}, or one of {nameof(Branch)} or {nameof(BranchParameter)} must be specified.");
+            }
+            
+            // Now it's GO TIME
+            Logger.Info($"Executing OctoVersion.Tool.");
+            var tempOutputFile = NukeBuild.TemporaryDirectory / $"octoversion.{Guid.NewGuid()}.json";
             var version = OctoVersionTasks.OctoVersionGetVersion(s => s
-                .DisableProcessLogOutput()
-                .SetOutputFormats(OctoVersionOutputFormatter.Json)
-                .SetFramework(Framework)
-                .When(NukeBuild.IsLocalBuild, s => s
-                    .SetCurrentBranch(GitTasks.GitCurrentBranch())))
+                    .SetFramework(Framework)
+                    .SetOutputJsonFile(tempOutputFile)
+                    .When(UpdateBuildNumber, x => x.EnableDetectEnvironment())
+                    .When(!UpdateBuildNumber, x => x.SetOutputFormats("JsonFile"))
+                    .When(!string.IsNullOrEmpty(octoversionBranch), x => x.SetCurrentBranch(octoversionBranch))
+                    .When(octoVersionMajor != null, x => x.SetMajor(octoVersionMajor))
```

Comment:
`When` is unnecessary here

Reply:
That implies a certain level of understanding of how the auto generated tool classes work.  
For anyone looking at the OctoVersion support and looking at this code here, this feels clearer that we dont pass these arguments to the tool when no value was resolved in the attribute.

If you strongly object I can remove the `.When` but I feel it does add clarity at the point where the tool is executed 🤣

Reply:
(i also had a look in the generated tool class and cant even see that it doesnt set the cmdline parameter when the passed value is null... so it must be even deeper in the execution flow if this does happen.  It feels alot clearer to me to leave these conditions at this obvious point where it sets up the tool execution

Reply:
Maybe I should communicate that better in the docs, but `SetX(null)` is essentially a `ResetX()`. It's probably important to understand that because it's supposed to avoid exactly this type of "excessive/repetitive/error-prone conditions".

Reply:
Ok I have removed all of those When clauses